fuel-indexer/indexer/src/database.rs

```diff
@@ -86,56 +73,73 @@ impl SchemaManager {
         let version = schema_version(schema);
 
         if !TypeIds::schema_exists(&*connection, name, &version)? {
-            let db_schema = SchemaBuilder::new(name, &version).build(schema);
-
-            connection.transaction::<(), diesel::result::Error, _>(|| {
-                for query in db_schema.statements.iter() {
-                    sql_query(query).execute(&*connection)?;
-                }
-
-                db_schema.commit_metadata(&*connection)?;
-
-                Ok(())
-            })?;
+            let _db_schema = SchemaBuilder::new(name, &version)
+                .build(schema)
+                .commit_metadata(&*connection)?;
         }
         Ok(())
     }
+
+    pub fn load_schema(&self, name: &str) -> IndexerResult<Schema> {
+        // TODO: might be nice to cache this data in server?
+        Ok(Schema::load_from_db(&*self.pool.get()?, name)?)
+    }
 }
 
 /// Database for an executor instance, with schema info.
-#[derive(Clone, Debug)]
+#[derive(Debug)]
 pub struct Database {
-    pub pool: DbPool,
+    pub conn: ConnWrapper,
     pub namespace: String,
     pub version: String,
     pub schema: HashMap<String, Vec<String>>,
     pub tables: HashMap<u64, String>,
 }
 
 impl Database {
-    pub fn new(db_conn: impl Into<String>) -> IndexerResult<Database> {
-        let pool = DbPool::new(db_conn)?;
+    pub fn new(db_conn: &str) -> IndexerResult<Database> {
+        let conn = ConnWrapper(PgConnection::establish(db_conn)?);
 
         Ok(Database {
-            pool,
+            conn,
             namespace: Default::default(),
             version: Default::default(),
             schema: Default::default(),
             tables: Default::default(),
         })
     }
 
-    fn upsert_query(&self, table: &str, inserts: Vec<String>) -> String {
+    pub fn start_transaction(&self) -> IndexerResult<usize> {
+        Ok(sql_query("BEGIN").execute(&*self.conn)?)
+    }
+
+    pub fn commit_transaction(&self) -> IndexerResult<usize> {
+        Ok(sql_query("COMMIT").execute(&*self.conn)?)
+    }
+
+    pub fn revert_transaction(&self) -> IndexerResult<usize> {
+        Ok(sql_query("ROLLBACK").execute(&*self.conn)?)
+    }
+
+    fn upsert_query(
+        &self,
+        table: &str,
+        columns: &[String],
+        inserts: Vec<String>,
+        updates: Vec<String>,
+    ) -> String {
         format!(
```

Comment:
Do we have sanitization concerns when constructing queries this way?

Reply:
Yeah, this one sort of has to be done like this. These are the tables generated by the user's graphql, so we don't have the luxury of auto-generated rust types here.